nerfstudio/models/kplanes.py

```diff
@@ -0,0 +1,396 @@
+            N = self.config.proposal_weights_anneal_max_num_iters
+
+            def set_anneal(step):
+                # https://arxiv.org/pdf/2111.12077.pdf eq. 18
+                train_frac = np.clip(step / N, 0, 1)
+                bias = lambda x, b: (b * x) / ((b - 1) * x + 1)
+                anneal = bias(train_frac, self.config.proposal_weights_anneal_slope)
+                self.proposal_sampler.set_anneal(anneal)
+
+            callbacks.append(
+                TrainingCallback(
+                    where_to_run=[TrainingCallbackLocation.BEFORE_TRAIN_ITERATION],
+                    update_every_num_iters=1,
+                    func=set_anneal,
+                )
+            )
+            callbacks.append(
+                TrainingCallback(
+                    where_to_run=[TrainingCallbackLocation.AFTER_TRAIN_ITERATION],
+                    update_every_num_iters=1,
+                    func=self.proposal_sampler.step_cb,
+                )
+            )
+        return callbacks
+
+    def get_outputs(self, ray_bundle: RayBundle):
+        ray_samples, weights_list, ray_samples_list = self.proposal_sampler(ray_bundle, density_fns=self.density_fns)
+        field_outputs = self.field(ray_samples)
+        weights = ray_samples.get_weights(field_outputs[FieldHeadNames.DENSITY])
+        weights_list.append(weights)
+        ray_samples_list.append(ray_samples)
+
+        if self.training and self.config.train_with_random_bg:
+            standard_bg_color = self.renderer_rgb.background_color
+            self.renderer_rgb.background_color = "random"
+
+        rgb = self.renderer_rgb(rgb=field_outputs[FieldHeadNames.RGB], weights=weights)
+
+        if self.config.train_with_random_bg:
+            rgb, bg_color = rgb
+
+        depth = self.renderer_depth(weights=weights, ray_samples=ray_samples)
+        accumulation = self.renderer_accumulation(weights=weights)
+
+        outputs = {
+            "rgb": rgb,
+            "accumulation": accumulation,
+            "depth": depth,
+        }
+
+        if self.training and self.config.train_with_random_bg:
+            self.renderer_rgb.background_color = standard_bg_color
+            outputs["bg_color"] = bg_color
+
+        # These use a lot of GPU memory, so we avoid storing them for eval.
+        if self.training:
+            outputs["weights_list"] = weights_list
+            outputs["ray_samples_list"] = ray_samples_list
+
+        for i in range(self.config.num_proposal_iterations):
+            outputs[f"prop_depth_{i}"] = self.renderer_depth(weights=weights_list[i], ray_samples=ray_samples_list[i])
+
+        return outputs
+
+    def get_metrics_dict(self, outputs, batch):
+        image = self._get_image(batch, outputs)
+
+        metrics_dict = {}
+        metrics_dict["psnr"] = self.psnr(outputs["rgb"], image)
+        if self.training:
+            metrics_dict["plane_tv"] = compute_grid_tv(self.field.grids)
```

Comment:
Maybe these should be computed in the `get_loss_dict` function directly (they're not really metrics like PSNR is).

Reply:
I personally like being able to look at these values before they get scaled with loss coefficients, but I don't have very strong opinions if others disagree